File: day05/day05.py
import time


with open("day05_input.txt", 'r') as day05_file:
    day05_lines = day05_file.read().splitlines()

seed_list = day05_lines[0].split()
seed_list.pop(0)
seed_list = [int(x) for x in seed_list]

remap_lists = []
new_remap_list = []
expect_a_new_list = False
for i in range(1, len(day05_lines)):
    if day05_lines[i] == "":
        expect_a_new_list = True
        if len(new_remap_list) > 0:
            remap_lists.append(new_remap_list)
        new_remap_list = []
        continue
    if expect_a_new_list:
        new_remap_list.append(day05_lines[i].split()[0])
        expect_a_new_list = False
        continue
    new_remap_element = []
    new_remap_split = day05_lines[i].split()
    for j in range(0, 3):
        new_remap_element.append(int(new_remap_split[j]))
    new_remap_list.append(new_remap_element)
if len(new_remap_list) > 0:
    remap_lists.append(new_remap_list)

# print(seed_list)
# for i in range(len(remap_lists)):
#     print(remap_lists[i])

# done collecting data, now for the actual solution...

def remap_number(initial_number, remap_element):
    for xI in range(1, len(remap_element)):
        if remap_element[xI][1] <= initial_number <= remap_element[xI][1] + remap_element[xI][2] - 1:
            return initial_number - (remap_element[xI][1] - remap_element[xI][0])
    return initial_number


phase01_minimum = ""
for i in range(len(seed_list)):
    seed_item = seed_list[i]
    # print(seed_item, end = " => ")
    for j in range(len(remap_lists)):
        seed_item = remap_number(seed_item, remap_lists[j])
        # print(seed_item, end=" => ")
    # print(" end ")
    if phase01_minimum == "":
        phase01_minimum = seed_item
    else:
        if phase01_minimum > seed_item:
            phase01_minimum = seed_item
print("Phase 01 result:", phase01_minimum)

# Phase 02. Tried to make a lazy solution, it worked too slowly... tried to make a smart solution, messed it up
# rolling back to the lazy solution since I'm past 24h+ mark already anyway

phase02_minimum = 10 ** 10
start_time = time.time()
counter = 0
for i in range(0, len(seed_list), 2):
    range_left = seed_list[i]
    range_right = seed_list[i] + seed_list[i + 1]
    for x in range(range_left, range_right):
        counter += 1
        if counter >= 1000000:
            new_time = time.time()
            print(new_time - start_time, "seconds passed. Processing", x, "in range", range_left, "-", range_right)
            counter = 0
        # print(seed_item, end = " => ")
        seed_item = x
        for j in range(len(remap_lists)):
            seed_item = remap_number(seed_item, remap_lists[j])
            # print(seed_item, end=" => ")
        # print(" end ")
        if phase02_minimum > seed_item:
            phase02_minimum = seed_item
print("Phase 02 result:", phase02_minimum)
f = open("output.txt", "w")
f.write("Phase 02 result: " + str(phase02_minimum))
f.close()

File: day05/day05.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"
)

func say(s string) {
	for i := 0; i < 5; i++ {
		time.Sleep(100 * time.Millisecond)
		fmt.Println(s)
	}
}

func main() {
	file, err := os.Open("day05_input.txt")
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()
	day05_lines := []string{}

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		day05_lines = append(day05_lines, scanner.Text())
	}
	// fmt.Println(day05_lines)

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	seedListString := strings.Fields(day05_lines[0])
	seedListString = seedListString[1:]
	var seedList []int
	for _, str := range seedListString {
		num, err := strconv.Atoi(str)
		if err != nil {
			fmt.Println("Error converting string to integer:", err)
			return
		}
		seedList = append(seedList, num)
	}
	// fmt.Println(seed_list)

	var remapLists [][]interface{}
	var newRemapList []interface{}
	expectANewList := false

	for i := 1; i < len(day05_lines); i++ {
		if day05_lines[i] == "" {
			expectANewList = true
			if len(newRemapList) > 0 {
				remapLists = append(remapLists, newRemapList)
			}
			newRemapList = nil
			continue
		}

		if expectANewList {
			splitResult := strings.Fields(day05_lines[i])
			if len(splitResult) > 0 {
				newRemapList = append(newRemapList, splitResult[0])
			}
			expectANewList = false
			continue
		}

		var newRemapElement []interface{}
		newRemapSplit := strings.Fields(day05_lines[i])
		for j := 0; j < 3; j++ {
			num, err := strconv.Atoi(newRemapSplit[j])
			if err == nil {
				newRemapElement = append(newRemapElement, num)
			} else {
				newRemapElement = append(newRemapElement, newRemapSplit[j])
			}
		}
		newRemapList = append(newRemapList, newRemapElement)
	}

	if len(newRemapList) > 0 {
		remapLists = append(remapLists, newRemapList)
	}

	// Print the resulting slice of slices
	// fmt.Println(remapLists)

	remapNumber := func(initialNumber int, remapElement []interface{}) int {
		for xI := 1; xI < len(remapElement); xI++ {
			switch v := remapElement[xI].(type) {
			case []interface{}:
				// If it's a []interface{}, perform the remapping logic
				if len(v) >= 3 {
					// Ensure that v has at least three elements before accessing v[1], v[2], etc.
					if start, ok := v[1].(int); ok {
						if end, ok := v[2].(int); ok && start <= initialNumber && initialNumber <= start+end-1 {
							return initialNumber - (start - v[0].(int))
						}
					}
				}
				// Add more cases if you have other types in remapElement
			}
			// case []int:
			// If it's a []int, perform the remapping logic
			// if v[1] <= initialNumber && initialNumber <= v[1]+v[2]-1 {
			// return initialNumber - (v[1] - v[0])
			// }
			// Add more cases if you have other types in remapElement
			// }
		}
		return initialNumber
	}

	phase02Minimum := 1000000000
	startTime := time.Now()

	processRange := func(wg *sync.WaitGroup, seedList []int, remapLists [][]interface{}, rangeLeft, rangeRight int, counter int) {
		defer wg.Done()

		for x := rangeLeft; x < rangeRight; x++ {
			counter++
			if counter >= 10000000 {
				newTime := time.Now()
				fmt.Printf("%v seconds passed. Processing %v in range %v - %v\n", newTime.Sub(startTime).Seconds(), x, rangeLeft, rangeRight)
				counter = 0
			}

			seedItem := x
			// fmt.Printf("Initial seed item: %v", seedItem)
			for j := 0; j < len(remapLists); j++ {
				seedItem = remapNumber(seedItem, remapLists[j])
				// fmt.Printf("New seed item: %v\n", seedItem)
			}

			if phase02Minimum > seedItem {
				phase02Minimum = seedItem
			}
		}
	}

	var wg sync.WaitGroup

	for i := 0; i < len(seedList); i += 2 {
		rangeLeft := seedList[i]
		rangeRight := seedList[i+1] + rangeLeft

		wg.Add(1)
		go processRange(&wg, seedList, remapLists, rangeLeft, rangeRight, 0)
	}

	wg.Wait()

	fmt.Printf("Phase 02 result: %v\n", phase02Minimum)
}
